File: src/util/Monitor.java
package util;

import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;

// Usage: Override getCurrent() method, then call start(), and call forceShutdown() if necessary.
public abstract class Monitor extends Thread {
    private String name;
    private int total, time;
    private PrintWriter out;
    private boolean stopped;
    private int current, lastTick;

    // Start from 0, end at <total>. <time> in seconds is the delay between 2 logs.
    // <total> == -1 indicates INF. In this case, we need to call forceShutdown() manually.
    public Monitor(String name, int total, int time, OutputStream out) {
        if (total < 0 && total != -1) {
            throw new RuntimeException("'total' is invalid.");
        }
        this.name = name;
        this.total = total;
        this.time = time;
        if (out == null) {
            this.out = null;
        } else {
            this.out = new PrintWriter(new OutputStreamWriter(out, StandardCharsets.UTF_8));
        }
        this.stopped = false;

        this.current = 0;
        this.lastTick = 0;
    }

    public Monitor(String name, int total, int time) {
        this(name, total, time, null);
    }

    public abstract int getCurrent();

    public final void log(String logString) {
        if (out == null) {
            System.out.println(logString);
        } else {
            out.println(logString);
            out.flush();
        }
    }

    public void setTotal(int total) {
        this.total = total;
    }

    // Override these log methods for another logging way.
    public void logStart() {
        log(String.format("MONITOR [%s] : STARTED.", name));
    }

    public void logDone() {
        log(String.format("MONITOR [%s] : DONE.", name));
    }

    public void logForcedShutdown() {
        log(String.format("MONITOR [%s] : FORCED SHUTDOWN.", name));
    }

    public void logProgress(Progress progress) {
        String currentString = progress.total == -1
                ? String.format("%d/--", progress.current)
                : String.format("%d/%d", progress.current, progress.total);
        String percentString = progress.total == -1 ? "--" : String.format("%.2f%%", progress.percent);
        String speedString = String.format("%.2f/sec", progress.speed);
        String etaString = progress.total == -1 ? "--d --:--:--" :
                String.format("%dd %02d:%02d:%02d", progress.eta_d, progress.eta_h, progress.eta_m, progress.eta_s);

        log(String.format("MONITOR [%s] : current: %s     percent: %s     speed: %s     eta: %s",
                name, currentString, percentString, speedString, etaString));
    }

    public void run() {
        logStart();
        if (total == 0) {
            stopped = true;
            logDone();
            return;
        }
        while (!stopped) {
            try {
                Thread.sleep(time * 1000);
            } catch (InterruptedException e) {
                break;
            }
            current = getCurrent();
            if (current < 0 || (total >= 0 && current > total)) {
                throw new RuntimeException("'getCurrent()' is invalid.");
            }

            // Compute progress.
            Progress progress = new Progress();
            progress.current = current;
            if (total != -1) {
                progress.total = total;
            }
            if (total != -1) {
                progress.percent = ((double) current) / total * 100;
            }
            progress.speed = ((double) current - lastTick) / time;
            if (Math.abs(progress.speed) >= 1e-2) {
                int remainingTimeInSeconds = (int) ((total - current) / progress.speed);
                progress.eta_d = remainingTimeInSeconds / (24 * 3600);
                progress.eta_h = (remainingTimeInSeconds % (24 * 3600)) / 3600;
                progress.eta_m = (remainingTimeInSeconds % 3600) / 60;
                progress.eta_s = remainingTimeInSeconds % 60;
            }
            logProgress(progress);

            lastTick = current;

            if (current == total) {
                stopped = true;
            }
        }
        current = getCurrent();
        if (current < 0 || (total >= 0 && current > total)) {
            throw new RuntimeException("'getCurrent()' is invalid.");
        }
        if (current == total) {
            logDone();
        } else {
            logForcedShutdown();
        }
    }

    public boolean forceShutdown() {
        if (!stopped) {
            stopped = true;
            this.interrupt();
            return true;
        } else {
            return false;
        }
    }

    @Override
    protected void finalize() throws Throwable {
        if (out != null) {
            out.close();
        }
        super.finalize();
    }

    public static final class Progress {
        int current = 0;
        int total = -1;
        double percent = -1; // -1 means unknown.
        double speed = 0;
        int eta_d = -1, eta_h = -1, eta_m = -1, eta_s = -1; // -1 means unknown.
    }
}

File: quid/quid/util/monitor.py
import sys
import time
from abc import abstractmethod, ABC
from threading import Thread, Event, Lock


class _Progress:
    def __init__(self):
        self.et_d = self.et_h = self.et_m = self.et_s = 0
        self.current = self.speed = 0
        self.total = self.percent = -1
        self.eta_d = self.eta_h = self.eta_m = self.eta_s = -1


class Monitor(ABC, Thread):
    def __init__(self, name=None, total=-1, period=10, out=sys.stdout):
        Thread.__init__(self)

        self._name = name
        self._total = total
        self._period = period
        self._out = out

        self._stop_event = Event()
        self._stopped = False

        self._last_tick = 0

    def set_total(self, total):
        self._total = total

    @abstractmethod
    def get_current(self):
        pass

    def _log(self, str):
        self._out.write(str)
        self._out.write('\n')

    def log_start(self):
        self._log('MONITOR [{}]: STARTED.'.format(self._name))

    def log_done(self):
        self._log('MONITOR [{}]: DONE'.format(self._name))

    def log_shutdown(self):
        self._log('MONITOR [{}]: SHUTDOWN.'.format(self._name))

    def log_progress(self, progress: _Progress):
        current_str = '{}/{}'.format(progress.current, '--' if progress.total == -1 else progress.total)
        percent_str = '' if progress.total == -1 else ' {:.2f}%'.format(progress.percent)
        speed_str = '{:.2f}/sec'.format(progress.speed)

        et_str = '{}d {:02d}:{:02d}:{:02d}'.format(progress.et_d, progress.et_h, progress.et_m, progress.et_s)
        eta_str = '--d --:--:--' if progress.total == -1 else \
            '{}d {:02d}:{:02d}:{:02d}'.format(progress.eta_d, progress.eta_h, progress.eta_m, progress.eta_s)
        self._log('MONITOR [{}]: et: {}    current: {}{}    speed: {}    eta: {}'.format(
            self.name, et_str, current_str, percent_str, speed_str, eta_str))

    def shutdown(self):
        if not self._stopped:
            self._stopped = True
            self._stop_event.set()
            return True
        else:
            return False

    def run(self):
        self.log_start()

        if self._total == 0:
            self._stopped = True
            self.log_done()
            return

        start_time = time.time()
        while not self._stopped:
            self._stop_event.wait(self._period)
            current = self.get_current()
            if current < 0 or 0 <= self._total < current:
                raise Exception('get_current() is invalid.')

            et_sec = int(time.time() - start_time)
            progress = _Progress()
            progress.et_d = int(et_sec / 24 / 3600)
            progress.et_h = int((et_sec % (24 * 3600)) / 3600)
            progress.et_m = int((et_sec % 3600) / 60)
            progress.et_s = int(et_sec % 60)
            progress.current = current

            if self._total != -1:
                progress.total = self._total
                progress.percent = current / self._total * 100

            progress.speed = (current - self._last_tick) / self._period

            if abs(progress.speed) >= 1e-2:
                rm_sec = int((self._total - current) / progress.speed)
                progress.eta_d = int(rm_sec / 24 / 3600)
                progress.eta_h = int((rm_sec % (24 * 3600)) / 3600)
                progress.eta_m = int((rm_sec % 3600) / 60)
                progress.eta_s = int(rm_sec % 60)

            self.log_progress(progress)

            self._last_tick = current

            if current == self._total:
                self._stopped = True

        current = self.get_current()

        if current < 0 or 0 <= self._total < current:
            raise Exception('get_current() is invalid.')

        if current == self._total:
            self.log_done()
        else:
            self.log_shutdown()


class CounterMonitor(Monitor):
    def __init__(self, name=None, total=-1, period=10, out=sys.stdout):
        super().__init__(name, total, period, out)
        self._curr = 0
        self._lock = Lock()

    def inc(self, count=1):
        with self._lock:
            self._curr += count
            return self._curr

    def get_current(self):
        return self._curr
